alloc.c

```diff
@@ -618,16 +618,17 @@ bool liftoff_output_apply(struct liftoff_output *output, drmModeAtomicReq *req)
 
 	/* Disable all planes. Do it before building mappings to make sure not
 	   to hit bandwidth limits because too many planes are enabled. */
+	liftoff_log_cnt(LIFTOFF_DEBUG, "Disabling planes:");
 	liftoff_list_for_each(plane, &device->planes, link) {
 		if (plane->layer == NULL) {
-			liftoff_log(LIFTOFF_DEBUG,
-				    "Disabling plane %"PRIu32, plane->id);
+			liftoff_log_cnt(LIFTOFF_DEBUG, " %"PRIu32, plane->id);
```

Comment:
Hmm. I'm on the fence about this, I'm not a fan of adding a new log API, especially given that some log APIs aren't able to control whether newlines are added to the output or not. For instance this is the case of wlroots' log system.

```c
if (log_has(LIFTOFF_DEBUG)) {
	char *buf = calloc(1, 11 * liftoff_list_length(&device->planes) + 1);
	size_t n = 0;
	liftoff_list_for_each(plane, &device->planes, link) {
		if (plane->layer == NULL) {
			int ret = snprintf(&buf[n], 12, " %"PRIu32, plane->id);
			if (ret < 0) {
				break;
			}
			n += ret - 1;
		}
	}
	liftoff_log(LIFTOFF_DEBUG, "Disabling planes:%s", buf);
	free(buf);
}

… // Proceed disabling the planes
```

Reply:
Having a buffer though whenever a line is not meant to be output directly is super complicated though. :(

Is there not a more stream-like way to do that?

What could be an alternative is to have in the log.c file a single buffer that fills with liftoff_log_cnt and is reallocated when it becomes full. Finally it is prepended whenever liftoff_log is called.

Reply:
I have uploaded a new patch where per-device log buffers are created for continuous debug messages. It's a bit more involved for what it's doing in the end but that way all memory management for debugging is in a single log function.

I have added the change as a new commit in the end but if you are ok with the solution I can also replace this earlier commit with it.